3131 errror in calculation of temperature margin bug

mkovari opened the pull request:
## Description

I have replaced the home-made Newton-Raphson solver and the secant solver with `newton` for calculating the temperature margin of superconductors.

## Checklist

I confirm that I have completed the following checks:

- [X] No large differences in the regression tests except for constraint residues.
- [X] No new tests
- [X] No doc changes

jonmaddock commented:
Hi Michael. Can you resolve the conflicts on this branch, then possibly tidy up the commit history with an interactive rebase if you feel that would make these changes clearer?

mkovari commented:
Jon,
I have had a look at the conflict but I don’t I understand what has caused it.  I have attempted to resolve it by simply restoring my code, but now the tests are failing when they weren’t yesterday.  I am afraid I don’t know what’s going on.

All the best,
Michael
From: jonmaddock ***@***.***>
Sent: Friday, April 19, 2024 9:04 AM
To: ukaea/PROCESS ***@***.***>
Cc: Kovari, Michael D ***@***.***>; Author ***@***.***>
Subject: Re: [ukaea/PROCESS] 3131 errror in calculation of temperature margin bug (PR #3142)


Hi Michael. Can you resolve the conflicts on this branch, then possibly tidy up the commit history with an interactive rebase if you feel that would make these changes clearer?

—
Reply to this email directly, view it on GitHub<https://github.com/ukaea/PROCESS/pull/3142#issuecomment-2066037517>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/ACKEVEAJZZQCLBFBXQQ7G5DY6DFX3AVCNFSM6AAAAABGLMN54CVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDANRWGAZTONJRG4>.
You are receiving this because you authored the thread.Message ID: ***@***.***>

mkovari commented:
Anyway - there are now no merge conflicts.
I don't know what an interactive rebase is.

mkovari commented:
I believe this is now ready to merge.   I did both the things we discussed:
1. Replaced `brentq` by `newton`. 
2. Restored the commented-out integration tests, changed the  value of `rohc` to something more sensible, and changed the two assertions that failed as a result.

jonmaddock commented on tests/integration/data/test_solver.conf:
```diff
@@ -2,7 +2,8 @@
 * No iterations
-NITER = 10
+# NITER = 10
+NITER = 1
```
What's the purpose of this change? The rest of the commit appears to be to do with getting the sctfcoil and superconductors tests to pass.

mkovari replied:
Reverted

jonmaddock commented on tests/integration/test_pfcoil_int.py:
```diff
@@ -35,241 +37,241 @@ def pfcoil():
+#     assert pytest.approx(pfv.rjohc) == -7.728453e9
+
+
+def test_efc(pfcoil: PFCoil[CsFatigue], monkeypatch: pytest.MonkeyPatch):
```
`pfcoil: PFCoil[CsFatigue]`: I'm not sure that's the correct type hint.

jonmaddock commented on tests/integration/test_pfcoil_int.py:
```diff
@@ -35,241 +37,241 @@ def pfcoil():
-
-
-def test_efc(pfcoil, monkeypatch):
+# def test_pfcoil(monkeypatch, pfcoil):
```
As all this is commented back in later, can this be fixed up to reduce churn in this PR?

jonmaddock commented on tests/integration/test_pfcoil_int.py:
```diff
@@ -271,7 +271,7 @@ def pfcoil():
 
 
-def test_efc(pfcoil: PFCoil[CsFatigue], monkeypatch: pytest.MonkeyPatch):
+def test_efc(pfcoil: PFCoil, monkeypatch: pytest.MonkeyPatch):
```
As this corrects an earlier error, can the previous commit be fixed up with this to reduce churn please?

mkovari replied:
Will do using rebase

jonmaddock commented on process/pfcoil.py:
```diff
@@ -2889,26 +2889,27 @@ def superconpf(
 
-            t_zero_margin, root_result = optimize.brentq(
+            another_estimate = 2 * thelium
+            t_zero_margin, root_result = optimize.newton(
```
As the solver has been changed from newton-raphson to brentq to newton, can this churn be fixed up please?

mkovari replied:
Will do using rebase

jonmaddock commented on tests/integration/test_pfcoil_int.py:
```diff
@@ -37,238 +37,238 @@ def pfcoil():
-
-#     assert pytest.approx(pfv.bpf[4]) == 9.299805e2
-#     assert pytest.approx(pfv.rjohc) == -7.728453e9
+def test_pfcoil(monkeypatch, pfcoil):
```
Commenting back in of commented-out function: please reduce churn.

mkovari replied:
Will do using rebase

jonmaddock commented on process/pfcoil.py:
```diff
@@ -2909,8 +2909,7 @@ def superconpf(
             )
-            print(root_result)  # Diagnostic for newton method
-            print("bmax = ", bmax)
+            # print(root_result)  # Diagnostic for newton method
```
Could be fixed-up.

mkovari replied:
Comment line removed

mkovari commented on process/maths_library.py:
```diff
@@ -4,66 +4,66 @@
 
-def secant_solve(f, x1, x2, opt_tol=None):
-    """Solve an equation f(x)=0.
+# def secant_solve(f, x1, x2, opt_tol=None):
```
Deleted

mkovari commented:
I think the conflict is resolved now.

jonmaddock commented:
I'm afraid there are still conflicts that need to be resolved. Please resolve them by rebasing, then re-request a review. Thanks.

mkovari commented:
GitHub says this branch has no conflicts with the base branch.  What conflicts are you thinking of?
